File: backend/src/dtos/common_dto.rs
use axum::{Json, http::StatusCode};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Debug, Serialize, Deserialize, ToSchema, Clone)]
pub struct CommonErrorResponse {
    pub error: String,
    pub code: u16,
}

impl CommonErrorResponse {
    pub fn new(error: String, code: StatusCode) -> Self {
        Self {
            error,
            code: code.as_u16(),
        }
    }

    pub fn to_response(&self) -> (StatusCode, Json<Self>) {
        let status = StatusCode::from_u16(self.code).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
        (status, Json(self.clone()))
    }
}

#[derive(Debug, Serialize, Deserialize, ToSchema, Clone)]
pub struct CommonResponse {
    pub message: String,
    pub code: u16,
}

impl CommonResponse {
    pub fn new(message: String, code: StatusCode) -> Self {
        Self {
            message,
            code: code.as_u16(),
        }
    }

    pub fn to_response(&self) -> (StatusCode, Json<Self>) {
        let status = StatusCode::from_u16(self.code).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
        (status, Json(self.clone()))
    }
}

#[derive(Debug, Deserialize)]
pub struct PaginationQuery {
    pub limit: Option<i64>,
    pub page: Option<i64>,
}

#[derive(Serialize, Clone, ToSchema)]
pub struct PaginatedResponse<T> {
    pub data: Vec<T>,
    pub page: i64,
    pub limit: i64,
    pub total: i64,
}

impl<T> PaginatedResponse<T>
where
    T: Serialize + Clone,
{
    pub fn new(data: Vec<T>, page: i64, limit: i64, total: i64) -> Self {
        Self {
            data,
            page,
            limit,
            total,
        }
    }

}

/// Simple response with only id and name for common listing endpoints
#[derive(Debug, Serialize, Deserialize, ToSchema, Clone)]
pub struct IdNameResponse {
    pub id: i64,
    pub name: String,
}

File: frontend/app/dtos/common.dto.ts
import { z } from 'zod'

export const CommonErrorResponseSchema = z.object({
  error: z.string(),
  code: z.number()
})

export type CommonErrorResponse = z.infer<typeof CommonErrorResponseSchema>

export const CommonResponseSchema = z.object({
  message: z.string(),
  code: z.number()
})

export type CommonResponse = z.infer<typeof CommonResponseSchema>

export const IdNameResponseSchema = z.object({
  id: z.number(),
  name: z.string()
})

export type IdNameResponse = z.infer<typeof IdNameResponseSchema>

export const createPaginatedResponseSchema = <T extends z.ZodTypeAny>(itemSchema: T) => z.object({
  data: z.array(itemSchema),
  page: z.number(),
  limit: z.number(),
  total: z.number()
})

export const PaginatedIdNameResponseSchema = createPaginatedResponseSchema(IdNameResponseSchema)

export type PaginatedIdNameResponse = z.infer<typeof PaginatedIdNameResponseSchema>

export const PaginationParamsSchema = z.object({
  page: z.number().optional(),
  limit: z.number().optional()
})

export type PaginationParams = z.infer<typeof PaginationParamsSchema>
